tests/js/server/replication2/replication2-replicated-log-wait-for-sync.js

```diff
@@ -0,0 +1,160 @@
+'use strict';
+
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2023 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License")
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Alexandru Petenchea
+////////////////////////////////////////////////////////////////////////////////
+
+const jsunity = require('jsunity');
+const arangodb = require("@arangodb");
+const db = arangodb.db;
+const helper = require('@arangodb/test-helper');
+const lh = require("@arangodb/testutils/replicated-logs-helper");
+const lp = require("@arangodb/testutils/replicated-logs-predicates");
+
+const database = "replication2_wait_for_sync_test_db";
+
+const replicatedLogSyncIndexSuiteReplication2 = function () {
+  const numberOfServers = 3;
+
+  const {setUpAll, tearDownAll, setUp, tearDownAnd} =
+    lh.testHelperFunctions(database, {replicationVersion: "2"});
+
+  const createReplicatedLog = function (database, config) {
+    const res = lh.createReplicatedLog(database, config, numberOfServers);
+    lh.waitFor(lp.replicatedLogIsReady(database, res.logId, res.term, res.servers, res.leader));
+    return res;
+  };
+
+  const clearAllFailurePoints = () => {
+    for (const server of lh.dbservers) {
+      helper.debugClearFailAt(lh.getServerUrl(server));
+    }
+  };
+
+  return {
+    setUpAll,
+    tearDownAll,
+    setUp,
+    tearDown: tearDownAnd(() => {
+      clearAllFailurePoints();
+    }),
+
+    testSyncIndexWaitForSyncTrue: function () {
+      const config = {
+        writeConcern: numberOfServers,
+        softWriteConcern: numberOfServers,
+        waitForSync: true,
+      };
+
+      const {logId} = createReplicatedLog(database, config);
+      let log = db._replicatedLog(logId);
+
+      // Insert 100 entries
+      for (let i = 0; i < 100; i++) {
+        log.ping();
+      }
+
+      // For wait-for-sync true, we expect the syncIndex to be equal to the spearhead
+      let status = log.status();
+      for (const [k, v] of Object.entries(status.participants)) {
+        assertEqual(v.response.local.syncIndex, v.response.local.spearhead.index,
+          `syncIndex should be equal to spearhead for ${k}, status: ${JSON.stringify(status)}`);
+        assertEqual(v.response.lowestIndexToKeep, 101,
+          `lowestIndexToKeep should be 101 for ${k}, status: ${JSON.stringify(status)}`);
+      }
```

Comment:
I was initially confused about this. The reason for my confusion was that I thought, the spearhead might be higher than the commit index. But `syncIndex` and `commitIndex` are not directly related. There is something like a `syncCommitIndex`, which is essentially calculating the commit index but using the `syncIndex` for each follower instead.

In `waitForSync=true` case they it is the same, however.

[This comment contains not action for you, please ignore it.]

Reply:
Added `syncCommitLeader` to `LeaderStatus`.